network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go

```diff
@@ -0,0 +1,180 @@
+// 3. The record is correctly initialized in the cache and can be retrieved using the GetWithInit method.
+func TestDuplicateMessageTrackerCache_ConcurrentSameRecordInit(t *testing.T) {
+	cache := duplicateMessageTrackerCacheFixture(t, 100, defaultDecay, defaultSkipDecayThreshold, zerolog.Nop(), metrics.NewNoopCollector())
+
+	peerID := unittest.PeerIdFixture(t)
+	const concurrentAttempts = 10
+
+	var wg sync.WaitGroup
+	wg.Add(concurrentAttempts)
+
+	for i := 0; i < concurrentAttempts; i++ {
+		go func() {
+			defer wg.Done()
+			gauge, found, err := cache.GetWithInit(peerID)
+			require.NoError(t, err)
+			require.True(t, found)
+			require.Zero(t, gauge)
+		}()
+	}
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 100*time.Millisecond, "timed out waiting for goroutines to finish")
+
+	// ensure that only one goroutine successfully initialized the record
+	require.Equal(t, uint(1), cache.c.Size())
+}
+
+// TestDuplicateMessageTrackerCache_DuplicateMessageReceived tests the DuplicateMessageReceived method of the RecordCache.
+// The test covers the following scenarios:
+// 1. Updating a record gauge for an existing peer ID.
+// 2. Attempting to update a record gauge  for a non-existing peer ID should not result in error. DuplicateMessageReceived should always attempt to initialize the gauge.
+// 3. Multiple updates on the same record only initialize the record once.
+func TestDuplicateMessageTrackerCache_DuplicateMessageReceived(t *testing.T) {
+	cache := duplicateMessageTrackerCacheFixture(t, 100, defaultDecay, defaultSkipDecayThreshold, zerolog.Nop(), metrics.NewNoopCollector())
+
+	peerID1 := unittest.PeerIdFixture(t)
+	peerID2 := unittest.PeerIdFixture(t)
+
+	gauge, err := cache.DuplicateMessageReceived(peerID1)
+	require.NoError(t, err)
+	require.Equal(t, float64(1), gauge)
+
+	// get will apply a slightl decay resulting
+	// in a gauge value less than gauge which is 1 but greater than 0.9
+	currentGauge, ok, err := cache.GetWithInit(peerID1)
+	require.NoError(t, err)
+	require.True(t, ok)
+	require.LessOrEqual(t, currentGauge, gauge)
+	require.Greater(t, currentGauge, 0.9)
+
+	_, ok, err = cache.GetWithInit(peerID2)
+	require.NoError(t, err)
+	require.True(t, ok)
+
+	// test adjusting the spam record for a non-existing node ID
+	peerID3 := unittest.PeerIdFixture(t)
+	gauge3, err := cache.DuplicateMessageReceived(peerID3)
+	require.NoError(t, err)
+	require.Equal(t, float64(1), gauge3)
+
+	// when updated the value should be incremented from 1 -> 2 and slightly decayed resulting
+	// in a gauge value less than 2 but greater than 1.9
+	gauge3, err = cache.DuplicateMessageReceived(peerID3)
+	require.NoError(t, err)
+	require.LessOrEqual(t, gauge3, 2.0)
+	require.Greater(t, gauge3, 1.9)
+}
+
+// TestDuplicateMessageTrackerCache_UpdateDecay ensures that a counter value in the record cache is eventually decayed back to 0 after some time.
+func TestDuplicateMessageTrackerCache_Decay(t *testing.T) {
```

Comment:
A concurrent version of this test is also imperative, i.e., calling `DuplicateMessageReceived` concurrently `x` times on the same `peerId` should result in the `GetWithInit` value of `> 1 and < x`.

Reply:
Case is covered in this test https://github.com/onflow/flow-go/pull/5058/commits/91a1a63836ca1b5a3e70065ec022fcc07397d03a